ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -387,11 +387,11 @@ short getVersion() {
    * @throws IOException
    */
   private void generateBloomFilterAndPersist() throws IOException {
-    List<IndexEntry> entries = new ArrayList<>();
-    getIndexEntriesSince(null, new FindEntriesCondition(Long.MAX_VALUE), entries, new AtomicLong(0), true);
-    bloomFilter = FilterFactory.getFilter(entries.size(), config.storeIndexBloomMaxFalsePositiveProbability);
-    for (IndexEntry entry : entries) {
-      bloomFilter.add(ByteBuffer.wrap(entry.getKey().toBytes()));
+    int numOfIndexEntries = numberOfEntries(mmap);
+    bloomFilter = FilterFactory.getFilter(numOfIndexEntries, config.storeIndexBloomMaxFalsePositiveProbability);
+    for (int i = 0; i < numOfIndexEntries; i++) {
+      StoreKey key = getKeyAt(mmap, i);
+      bloomFilter.add(ByteBuffer.wrap(key.toBytes()));
```

Comment:
what happens if you add the same key to the bloom filter multiple times? Is that ok? If not, we might need a `prevKey` and add only if `!key.equals(prevKey)`

Reply:
Adding the same key multiple times should have no effect as the bit positions for a key is fixed (so it would essentially set the key's hashed bits to 1 when they were already set to 1).

The other option is to simply check if the key is already present in the filter and only add it if it isn't (like the store does today when deleting) - but I don't think it is necessary.